cpp/src/arrow/flight/sql/server_session_middleware.cc

```diff
@@ -0,0 +1,207 @@
+                   std::shared_ptr<ServerMiddleware>* middleware);
+
+  /// \brief Get a new, empty session option map and its id key.
+  std::pair<std::string, std::shared_ptr<FlightSqlSession>> GetNewSession();
+};
+
+class ServerSessionMiddlewareImpl : public ServerSessionMiddleware {
+ protected:
+  std::shared_mutex lock_;
+  ServerSessionMiddlewareFactory* factory_;
+  const CallHeaders& headers_;
+  std::shared_ptr<FlightSqlSession> session_;
+  std::string session_id_;
+  const bool existing_session;
+
+ public:
+  ServerSessionMiddlewareImpl(ServerSessionMiddlewareFactory* factory,
+                              const CallHeaders& headers)
+      : factory_(factory), headers_(headers), existing_session(false) {}
+
+  ServerSessionMiddlewareImpl(ServerSessionMiddlewareFactory* factory,
+                              const CallHeaders& headers,
+                              std::shared_ptr<FlightSqlSession> session,
+                              std::string session_id)
+      : factory_(factory),
+        headers_(headers),
+        session_(std::move(session)),
+        session_id_(std::move(session_id)),
+        existing_session(true) {}
+
+  void SendingHeaders(AddCallHeaders* add_call_headers) override {
+    if (!existing_session && session_) {
+      add_call_headers->AddHeader(
+          "set-cookie", static_cast<std::string>(kSessionCookieName) + "=" + session_id_);
+    }
+  }
+
+  void CallCompleted(const Status&) override {}
+
+  bool HasSession() const override { return static_cast<bool>(session_); }
+
+  std::shared_ptr<FlightSqlSession> GetSession() override {
+    const std::shared_lock<std::shared_mutex> l(lock_);
+    if (!session_) {
+      auto [id, s] = factory_->GetNewSession();
+      session_ = std::move(s);
+      session_id_ = std::move(id);
+    }
+    return session_;
+  }
+
+  const CallHeaders& GetCallHeaders() const override { return headers_; }
+};
+
+std::vector<std::pair<std::string, std::string>>
+ServerSessionMiddlewareFactory::ParseCookieString(const std::string_view& s) {
+  const std::string list_sep = "; ";
+  const std::string pair_sep = "=";
+  const size_t pair_sep_len = pair_sep.length();
+
+  std::vector<std::pair<std::string, std::string>> result;
+
+  size_t cur = 0;
+  while (cur < s.length()) {
+    const size_t end = s.find(list_sep, cur);
+    size_t len;
+    if (end == std::string::npos) {
+      // No (further) list delimiters
+      len = std::string::npos;
+      cur = s.length();
+    } else {
+      len = end - cur;
+      cur = end;
+    }
+    const std::string_view tok = s.substr(cur, len);
+
+    const size_t val_pos = tok.find(pair_sep);
+    if (val_pos == std::string::npos) {
+      // The cookie header is somewhat malformed; ignore the key and continue parsing
+      continue;
+    }
+    result.emplace_back(tok.substr(0, val_pos),
+                        tok.substr(val_pos + pair_sep_len, std::string::npos));
+  }
```

Comment:
I discussed this fairly extensively with @lidavidm and the existing cookie functionality is heavily intertwined with managing the client-side cookie jar, so essentially we would need to factor the zero-copy stateful parser out as a generator which is getting pretty involved and could accidentally break the existing code.

Reply:
OK. But the implementation is buggy:

* `a=b` returns an empty vector instead of `[<a,b>]`
* `a=b;c=d` returns an empty vector instead of `[<a,b>, <c,d>]`
* `a=b; c=d` causes an infinite loop

Can we add tests?

Reply:
On that note, circling back around on tests, is there a trick to getting ARROW_BUILD_TESTS to not explode?  Getting a massive pile of errors, topmost of which is:

`gmock-actions.h:342:5: error: unknown type name 'GTEST_DISALLOW_COPY_AND_ASSIGN_'`

@kou @lidavidm ?  (C++ dev docs seem to suggest it Just Works)

Reply:
Could you show full error log?

Reply:
```
$ cmake --build .
[19/400] Building CXX object _deps/googletest-build/googlemock/CMakeFiles/gmock.dir/src/gmock-all.cc.o
FAILED: _deps/googletest-build/googlemock/CMakeFiles/gmock.dir/src/gmock-all.cc.o 
/opt/homebrew/bin/ccache /Library/Developer/CommandLineTools/usr/bin/c++ -DARROW_EXTRA_ERROR_CONTEXT -DARROW_HAVE_NEON -DARROW_WITH_TIMING_TESTS -DGTEST_CREATE_SHARED_LIBRARY=1 -Dgmock_EXPORTS -I/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include -I/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock -isystem /Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googletest/include -isystem /Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googletest -Qunused-arguments -fcolor-diagnostics -Wno-unused-value -Wno-ignored-attributes -g -Werror -O0 -ggdb  -Wno-error -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk -fPIC -Wall -Wshadow -Werror -Wconversion -DGTEST_HAS_PTHREAD=1 -fexceptions -W -Wpointer-arith -Wreturn-type -Wcast-qual -Wwrite-strings -Wswitch -Wunused-parameter -Wcast-align -Wchar-subscripts -Winline -Wredundant-decls -std=c++17 -MD -MT _deps/googletest-build/googlemock/CMakeFiles/gmock.dir/src/gmock-all.cc.o -MF _deps/googletest-build/googlemock/CMakeFiles/gmock.dir/src/gmock-all.cc.o.d -o _deps/googletest-build/googlemock/CMakeFiles/gmock.dir/src/gmock-all.cc.o -c /Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/src/gmock-all.cc
In file included from /Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/src/gmock-all.cc:39:
In file included from /Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock.h:59:
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-actions.h:342:5: error: unknown type name 'GTEST_DISALLOW_COPY_AND_ASSIGN_'
    GTEST_DISALLOW_COPY_AND_ASSIGN_(FixedValueProducer);
    ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-actions.h:353:5: error: unknown type name 'GTEST_DISALLOW_COPY_AND_ASSIGN_'
    GTEST_DISALLOW_COPY_AND_ASSIGN_(FactoryValueProducer);
    ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-actions.h:427:3: error: unknown type name 'GTEST_DISALLOW_COPY_AND_ASSIGN_'
  GTEST_DISALLOW_COPY_AND_ASSIGN_(ActionInterface);
  ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-actions.h:687:27: error: expected parameter declarator
    GTEST_COMPILE_ASSERT_(!std::is_reference<Result>::value,
                          ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-actions.h:687:27: error: expected ')'
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-actions.h:687:26: note: to match this '('
    GTEST_COMPILE_ASSERT_(!std::is_reference<Result>::value,
                         ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-actions.h:687:5: error: a type specifier is required for all declarations
    GTEST_COMPILE_ASSERT_(!std::is_reference<Result>::value,
    ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-actions.h:694:5: error: unknown type name 'GTEST_DISALLOW_COPY_AND_ASSIGN_'
    GTEST_DISALLOW_COPY_AND_ASSIGN_(Impl);
    ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-actions.h:659:9: error: use of undeclared identifier 'use_ReturnRef_instead_of_Return_to_return_a_reference'
        use_ReturnRef_instead_of_Return_to_return_a_reference);
        ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-actions.h:763:27: error: use of undeclared identifier 'use_Return_instead_of_ReturnRef_to_return_a_value'
                          use_Return_instead_of_ReturnRef_to_return_a_value);
                          ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-actions.h:806:9: error: use of undeclared identifier 'use_Return_instead_of_ReturnRefOfCopy_to_return_a_value'
        use_Return_instead_of_ReturnRefOfCopy_to_return_a_value);
        ^
In file included from /Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/src/gmock-all.cc:39:
In file included from /Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock.h:61:
In file included from /Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-function-mocker.h:42:
In file included from /Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-spec-builders.h:75:
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-matchers.h:316:3: error: unknown type name 'GTEST_DISALLOW_COPY_AND_ASSIGN_'
  GTEST_DISALLOW_COPY_AND_ASSIGN_(StringMatchResultListener);
  ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-matchers.h:538:7: error: use of undeclared identifier 'cannot_convert_non_reference_arg_to_reference'
      cannot_convert_non_reference_arg_to_reference);
      ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-matchers.h:548:7: error: use of undeclared identifier 'conversion_of_arithmetic_types_must_be_lossless'
      conversion_of_arithmetic_types_must_be_lossless);
      ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-matchers.h:683:25: error: use of undeclared identifier 'matcher_and_value_have_different_numbers_of_fields'
                        matcher_and_value_have_different_numbers_of_fields);
                        ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-matchers.h:2490:5: error: unknown type name 'GTEST_DISALLOW_COPY_AND_ASSIGN_'
    GTEST_DISALLOW_COPY_AND_ASSIGN_(Impl);
    ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-matchers.h:2505:7: error: expected parameter declarator
      !IsHashTable<GTEST_REMOVE_REFERENCE_AND_CONST_(RhsContainer)>::value,
      ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-matchers.h:2505:7: error: expected ')'
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-matchers.h:2504:24: note: to match this '('
  GTEST_COMPILE_ASSERT_(
                       ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-matchers.h:2504:3: error: a type specifier is required for all declarations
  GTEST_COMPILE_ASSERT_(
  ^
/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googlemock/include/gmock/gmock-matchers.h:2527:9: error: use of undeclared identifier 'use_UnorderedPointwise_with_hash_tables'
        use_UnorderedPointwise_with_hash_tables);
        ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
[30/400] Building CXX object _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o
ninja: build stopped: subcommand failed.
```

Reply:
Could you confirm whether `/Users/pauln/git/arrow/cpp/session-management-build-debug/_deps/googletest-src/googletest/include/gtest/internal/gtest-port.h` is included or not? (For example, we can add `#error XXX` to the `gtest-port.h`.)
I think that the `gtest-port.h` isn't included.

Reply:
I do now see the error directive when building:
```
[44/400] Building CXX object _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o
```

However just the above-pasted errors on the step for _deps/googletest-build/googlemock/CMakeFiles/gmock.dir/src/gmock-all.cc.o, so it's included in some places, but apparently not there as you figured.

Reply:
(Probably fixed the original bug—I see what I did when I did it; tests forthcoming once I get tests working!)

Reply:
Ping, any further thoughts on getting tests working locally?  Are you & David developing on Linux or where is this working properly?

Reply:
I can run tests on Linux and macOS.
Could you provide full `cmake ...` log and `cmake --build ...` log (with a clean build directory)?